### ISSUE TITLE
L7 redirects are not properly deleted

### ISSUE THREAD

**Issue by rlenglet** [cilium/cilium#2796]
We see regularly such warning messages in the logs:
```
Feb 12 16:18:05 cilium-master cilium-agent[11672]: level=warning msg="Error while removing proxy redirect" PolicyID.L4="38939:ingress:TCP:80" containerID=2bf25477f9 endpointID=38939 error="unable to find redirect 38939:ingress:TCP:80" policyRevision=33
Feb 12 16:18:05 cilium-master cilium-agent[11672]: level=warning msg="Error while removing proxy redirect" PolicyID.L4="38939:ingress:TCP:8080" containerID=2bf25477f9 endpointID=38939 error="unable to find redirect 38939:ingress:TCP:8080" policyRevision=33
```

Those messages are logged in `Proxy.RemoveRedirect` (`pkg/proxy/proxy.go`), when attempting to delete a non-existent redirect. This should not happen when Cilium is healthy, but we still see those messages regularly.

This is a symptom that Cilium is not properly deleting redirects, and may leave orphan redirect ports open in proxies.

The simplified structure of the code is:
```
func (e *Endpoint) regenerate(...) {
	...
	e.removeCollectedRedirects(...)
	...
	e.regenerateBPF(...)
	...
}

func (e *Endpoint) removeCollectedRedirects(...) {
	for id := range e.proxiesToRemove {
		... owner.RemoveProxyRedirect(..., id) ...
		delete(e.proxiesToRemove, id)
	}
}

func (d *Daemon) RemoveProxyRedirect(...) {
	...
	... d.l7proxy.RemoveRedirect(...)
}

func (e *Endpoint) regenerateBPF(...) {
	...
	e.regeneratePolicy(...)
	...
}

func (e *Endpoint) regeneratePolicy(...) {
	...
	e.regenerateConsumable(..)
	...
}

func (e *Endpoint) regenerateConsumable(...) {
	...
	e.collectUnusedRedirects(e.L4Policy.Ingress, c.L4Policy.Ingress)
	e.collectUnusedRedirects(e.L4Policy.Egress, c.L4Policy.Egress)
	...
}

func (e *Endpoint) collectUnusedRedirects(...) {
	...
	e.proxiesToRemove = make(map[string]bool)
	...
	e.proxiesToRemove[e.ProxyID(&v)] = true
	...
}
```

In summary, we try to delete redirects (call `removeCollectedRedirects`) before we've even calculated the set of redirects to remove. The accesses to `Endpoint.proxiesToRemove` are in the wrong order.

The solution would be to move the call to `removeCollectedRedirects` into `regenerateBPF`, after the call to `regeneratePolicy`.

**Comment by rlenglet** [comment]
@tgraf Could that be the cause of the leftover redirects issue we've hit recently?

**Comment by rlenglet** [comment]
Moving the call to `removeCollectedRedirects` into `regenerateBPF` after the call to `regeneratePolicy` improved the situation, but that was not enough. I still saw the same warning messages, but mostly preceded by `Skipping policy recalculation` log messages which indicated that `Endpoint.regeneratePolicy` skipped the call to `Endpoint.regenerateConsumable`, and therefore the `Endpoint.proxiesToRemove` set was still not populated or was stale in some cases. Always clearing the `Endpoint.proxiesToRemove` set in `regenerateBPF` did the trick.

**Comment by eloycoto** [comment]
Hi @rlenglet 

When you kill a container, I think that delete a redirect happens two times, one on stop event, and another one in delete event. Maybe it's related to that, but I'm not sure, I saw that last week too in issue #2707

**Comment by rlenglet** [comment]
@eloycoto no, endpoint delete was fine. That code is much simpler and works.
The problem I saw was with policy updates and regeneration.
